department-of-veterans-affairs/vets-api
# Handle Breakers::OutageException in Post911GIBillStatusesController

The service which Lighthouse uses to request GI Bill status information for a veteran occasionally experiences outages.  Vets API uses the Breakers gem to better handle outages by raising the OutageException class.  The ApplicationController is configured to raise a `ServiceOutage` error in these cases.

This commit adds handling for `Breakers::OutageException` in the `Post911GIBillStatusesController`'s show method, and to the associated service class which makes a request of Lighthouse.  When a `Breakers::OutageException` is encountered, it is simply raised, allowing the `ApplicationController` to handle it.

The `ServiceOutage` class was modified to include a status code of 503 in the data passed to `SerializableError`. This status code represents "Service Unavailable", which is the appropriate code to use when a service outage occurs.

Tests have been updated to cover this new exception handling logic and ensure that the correct HTTP status code and error message are returned when a `Breakers::OutageException` is raised.

## Related issue(s)

https://github.com/department-of-veterans-affairs/va-iir/issues/828

## Testing done

- Previously a 500 error would be raised, because a `Breakers::OutageException` does not have an `errors` attribute
- To test:
    - Pull the code in the ~[vets-api-mockdata PR](https://github.com/department-of-veterans-affairs/vets-api-mockdata/pull/498) down (assuming it's not been merged into main yet)~ MERGED! Just pull the main branch
    - Update your `settings.local.yml`: 
  
```
lighthouse:
  benefits_education:
    use_mocks: true
betamocks:
  enabled: true
  recording: false
  cache_dir: "../vets-api-mockdata"
```


    - Log in as `vets.gov.user+10@gmail.com` 
    - Go to http://localhost:3001/education/gi-bill/post-9-11/ch-33-benefit/status/
    - You should see the "system is down" message.

## What areas of the site does it impact?
- Modify [ServiceOutage](https://github.com/department-of-veterans-affairs/vets-api/blob/704f31094a8815062eb2236bfeed79eb4a4a759e/lib/common/exceptions/service_outage.rb) object: include `status: 503` in the error
- Modifications to [Post911GiBillStatusesController](https://github.com/department-of-veterans-affairs/vets-api/blob/704f31094a8815062eb2236bfeed79eb4a4a759e/app/controllers/v1/post911_gi_bill_statuses_controller.rb) and [BenefitsEducation::Service](https://github.com/department-of-veterans-affairs/vets-api/blob/704f31094a8815062eb2236bfeed79eb4a4a759e/lib/lighthouse/benefits_education/service.rb) to allow `Breakers::OutageException` to bubble up and be handled by [ApplicationController](https://github.com/department-of-veterans-affairs/vets-api/blob/f9230a47d7887424b0b16c4de3c62f9ce2b58956/app/controllers/application_controller.rb) (included in [ExceptionHandling](https://github.com/department-of-veterans-affairs/vets-api/blob/710119ac09e28f9bfe43fed7467711c499afa094/app/controllers/concerns/exception_handling.rb))

## Acceptance criteria
